File: mgm-gcp/src/main/java/uk/co/cerihughes/mgm/translate/spotify/SpotifyAlbumTranslation.java
package uk.co.cerihughes.mgm.translate.spotify;

import com.wrapper.spotify.SpotifyApi;
import com.wrapper.spotify.model_objects.specification.Album;
import com.wrapper.spotify.model_objects.specification.ArtistSimplified;
import org.openapitools.model.AlbumApiModel;
import uk.co.cerihughes.mgm.model.interim.InterimAlbum;
import uk.co.cerihughes.mgm.model.interim.InterimEvent;
import uk.co.cerihughes.mgm.translate.AlbumTranslation;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

public final class SpotifyAlbumTranslation extends SpotifyTranslation implements AlbumTranslation {
    private final SpotifyApi spotifyApi;

    private final GetAlbumsOperation getAlbumsOperation = new GetAlbumsOperation();
    private Map<String, Album> preprocessedAlbums = new HashMap<>();

    public SpotifyAlbumTranslation(SpotifyApi spotifyApi) {
        super();

        this.spotifyApi = spotifyApi;
    }

    @Override
    public void preprocess(List<InterimEvent> interimEvents) {
        if (interimEvents == null || interimEvents.size() == 0) {
            return;
        }

        preprocessAlbums(interimEvents);
    }

    protected void preprocessAlbums(List<InterimEvent> interimEvents) {
        final ArrayList<InterimAlbum> interimAlbums = new ArrayList<>();
        for (InterimEvent interimEvent : interimEvents) {
            InterimAlbum classicAlbum = interimEvent.getClassicAlbum();
            InterimAlbum newAlbum = interimEvent.getNewAlbum();
            if (classicAlbum != null) {
                interimAlbums.add(classicAlbum);
            }
            if (newAlbum != null) {
                interimAlbums.add(newAlbum);
            }

        }
        final List<String> albumIds = interimAlbums.stream()
                .map(InterimAlbum::getAlbumData)
                .filter(s -> isValidData(s))
                .collect(Collectors.toList());
        final List<Album> albums = getAlbumsOperation.execute(spotifyApi, albumIds);
        preprocessedAlbums = albums.stream()
                .collect(Collectors.toMap(Album::getId, Function.identity()));
    }

    @Override
    public AlbumApiModel translate(InterimAlbum interimAlbum) {
        final String spotifyId = interimAlbum.getAlbumData();
        if (isValidData(spotifyId) == false) {
            return null;
        }

        final AlbumApiModel.TypeEnum type = interimAlbum.getType();
        if (type == null || spotifyId == null) {
            return null;
        }

        final Album spotifyAlbum = preprocessedAlbums.get(spotifyId);
        if (spotifyAlbum == null) {
            return null;
        }

        final ArtistSimplified[] spotifyArtists = spotifyAlbum.getArtists();
        if (spotifyArtists == null || spotifyArtists.length == 0) {
            return null;
        }

        final ArtistSimplified spotifyArtist = spotifyArtists[0];
        final String name = spotifyAlbum.getName();
        final String artist = spotifyArtist.getName();
        final AlbumApiModel album = new AlbumApiModel();
        album.setType(type);
        album.setSpotifyId(spotifyId);
        album.setName(name);
        album.setArtist(artist);
        album.setScore(interimAlbum.getScore());
        album.setImages(getImages(spotifyAlbum.getImages()));
        return album;
    }
}

File: mgm-gcp/src/main/java/uk/co/cerihughes/mgm/translate/spotify/SpotifyAlbumTranslation.kt
package uk.co.cerihughes.mgm.translate.spotify

import com.wrapper.spotify.SpotifyApi
import com.wrapper.spotify.model_objects.specification.Album
import uk.co.cerihughes.mgm.model.interim.InterimAlbum
import uk.co.cerihughes.mgm.model.interim.InterimEvent
import uk.co.cerihughes.mgm.model.output.AlbumApiModel
import uk.co.cerihughes.mgm.translate.AlbumTranslation

class SpotifyAlbumTranslation(private val spotifyApi: SpotifyApi) : SpotifyTranslation(), AlbumTranslation {
    private val getAlbumsOperation = GetAlbumsOperation()
    private var preprocessedAlbums: Map<String, Album> = HashMap()
    override fun preprocess(interimEvents: List<InterimEvent>) {
        preprocessAlbums(interimEvents)
    }

    protected fun preprocessAlbums(interimEvents: List<InterimEvent>) {
        val interimAlbums = interimEvents
            .flatMap { listOf(it.classicAlbum, it.newAlbum) }
        val albumIds = interimAlbums
            .map { it.albumData }
            .filter { isValidData(it) }
        val albums = getAlbumsOperation.execute(spotifyApi, albumIds)
        preprocessedAlbums = albums.map { it.id to it }
            .toMap()
    }

    override fun translate(interimAlbum: InterimAlbum): AlbumApiModel? {
        val spotifyId = interimAlbum.albumData
        if (!isValidData(spotifyId)) {
            return null
        }

        val spotifyAlbum = preprocessedAlbums[spotifyId] ?: return null
        val spotifyArtists = spotifyAlbum.artists
        if (spotifyArtists == null || spotifyArtists.size == 0) {
            return null
        }
        val spotifyArtist = spotifyArtists[0]
        val name = spotifyAlbum.name
        val artist = spotifyArtist.name
        val album = AlbumApiModel()
        album.type = interimAlbum.type
        album.spotifyId = spotifyId
        album.name = name
        album.artist = artist
        album.score = interimAlbum.score
        album.images = getImages(spotifyAlbum.images)
        return album
    }
}
